pages/_app.tsx

```diff
@@ -68,7 +68,32 @@ interface dataLayerItem {
 declare global {
   var dataLayer: dataLayerItem[]; // eslint-disable-line no-var
 }
+const useIsEngaged = () => {
+  const router = useRouter();
+  const [timerReached, setTimerReached] = useState(false);
+  const [secondPageReached, setSecondPageReached] = useState(false);
+  const [isEngaged, setIsEngaged] = useState(false);
+
+  useEffect(() => {
+    setTimeout(() => {
+      setTimerReached(true);
+    }, 30000);
+    const routeChanged = () => {
+      setSecondPageReached(true);
+    };
+
+    router.events.on("routeChangeComplete", routeChanged);
+    return () => {
+      router.events.off("routeChangeComplete", routeChanged);
+    };
+  }, [router.events]);
+
+  useEffect(() => {
+    setIsEngaged(secondPageReached && timerReached);
+  }, [secondPageReached, timerReached]);
```

Comment:
does this have to be two effects? From what I can tell from the logic, we are checking
1. if the second page was reached
2. if they've been here for 30 seconds

Since the timer isn't being reset anywhere, we can just put `setIsEngaged` into the timeout instead right? this removes extra state and an extra useEffect. something like this pseudo code below. what do you think?

```js
  useEffect(() => {
    const timeout = setTimeout(() => {
      setIsEngaged(secondPageReached);
    }, 30000);

    const routeChanged = () => {
      setSecondPageReached(true);
    };

    router.events.on('routeChangeComplete', routeChanged);
    return () => {
      clearTimeout(timeout);
      router.events.off('routeChangeComplete', routeChanged);
    };
  }, [router.events, secondPageReached]);
```

Reply:
Doesn't quite match the criteria for tracking the engagement metrics. 

With the current logic the timeout continues through page navigation and will trigger the tracking events when

1. The User has spent 30 seconds on the site
2. The user has visited at least two pages

With this proposed implementation the timeout resets on page navigation and the user would have to spend the full 30 seconds on the second (or any subsequent) page visited to be marked as an engaged viewer. This would cause a rapidly navigating user to not be marked as an engaged viewer.